### PR TITLE
Refactor: Minimize params size

### PR DESCRIPTION
**Pull request by fabianschu**
DONE: 
* simplified inputSchema and parameters that the deploy function takes:
  * no more array structures so that sdk user doesn't have to care about order of params
  * no more mention of `optionalModules` in any user facing data => for a user every module is treated the same way
  * no more mention of moduleTypes in any user facing data => a user requests eg `RebasingFundingManager`, `RoleAuthorizer`, `RecurringPaymentManager` and these module names will then also be the identifying keys that the user will have to work with (and not anymore *authorizer*, *paymentProcessor*, *paymentManager*)
* the user requests modules same as before:
```
[
      { name: 'RebasingFundingManager', version: 'v1.0' },
      { name: 'RoleAuthorizer', version: 'v1.0' },
      { name: 'SimplePaymentProcessor', version: 'v1.0' },
      { name: 'RecurringPaymentManager', version: 'v1.0' },
]
```
* the `inputSchema` that they get returned now looks as follows (note: RecurringPaymentManager is an optional module here):

```
{
  Orchestrator: {
    version: "v1.0",
    params: {
      owner: [Object ...],
      token: [Object ...],
    },
  },
  RebasingFundingManager: {
    version: "v1.0",
    params: {
      orchestratorTokenAddress: [Object ...],
    },
  },
  RoleAuthorizer: {
    version: "v1.0",
    params: {
      initialOwner: [Object ...],
      initialManager: [Object ...],
    },
  },
  RecurringPaymentManager: {
    version: "v1.0",
    params: {
      epochLength: [Object ...],
    },
  },
}
```
* where each `params` looks as follows:
```
{
  epochLength: {
    type: "uint256",
    description: "The length of an epoch in seconds. This will be the common denominator for all payments, as these are specified in epochs (i.e. if an epoch is 1 week, vestings can be done for 1 week, 2 week, 3 week, etc.). Epoch needs to be greater than 1 week and smaller than 52 weeks",
    jsType: "number",
  },
}
```
* the args object that the user needs to input into the deploy function now looks as follows:
```
{
  Orchestrator: {
    owner: "0x7AcaF5360474b8E40f619770c7e8803cf3ED1053",
    token: "0x5eb14c2e7D0cD925327d74ae4ce3fC692ff8ABEF",
  },
  RebasingFundingManager: {
    orchestratorTokenAddress: "0x5eb14c2e7D0cD925327d74ae4ce3fC692ff8ABEF",
  },
  RoleAuthorizer: {
    initialOwner: "0x7AcaF5360474b8E40f619770c7e8803cf3ED1053",
    initialManager: "0x7AcaF5360474b8E40f619770c7e8803cf3ED1053",
  },
  RecurringPaymentManager: {
    epochLength: "604800",
  },
}
```

### CODE REVIEW

File: tests/getDeploy.test.ts
Comment (0xxmemo) [inline tests/getDeploy.test.ts:1]: It is very hard to understand what is going on in this test, maybe have multiple tests for multiple cases

File: src/getDeploy/index.ts [hunk 1]
```diff
@@ -1,175 +1,181 @@
...
+    params: jsParams,
   }
-  return moduleSchema
+  return { moduleSchema, moduleName: name }
 }
 
 // based on the module names and versions passed to it
 // retrieves from the abi config the required deployment inputs
 // for the requested modules
 const getInputSchema = (modules: ModuleSpec[]) => {
```
Comment (0xxmemo) [inline src/getDeploy/index.ts]: this should be a ModuleSpec Object with keys as labels in this format =>

```
{
orchestrator: ModuleSpec
paymentProcessor: ModuleSpec
fundingManager: ModuleSpec
logicModule?: ModuleSpec[]
}
```

File: src/getDeploy/types.ts [hunk 2]
```diff
@@ -51,6 +51,11 @@ export type Params = {
   value: string
 }
 
+export type SmallParams = {
+  type: AbiType
+  value: string
+}
+
 export type ModuleSchema = {
```
Comment (0xxmemo) [inline src/getDeploy/types.ts:98]: Tbh we should just use FormattedAbiParameter from src/types/parameter

File: src/getDeploy/index.ts [hunk 3]
```diff
@@ -1,175 +1,181 @@
...
-} from './types'
-import { assembleMetadata, getDeploymentArgs, getWriteFn } from './utils'
+import { encodeAbiParameters } from 'viem'
+import { ORCHESTRATOR_CONFIG } from './constants'
+import { ModuleSpec, UserInputs } from './types'
+import { assembleMetadata, getDeploymentConfig, getWriteFn } from './utils'
+import { getModuleVersion } from '@inverter-network/abis'
 
 // creates an array of length = 5 to fill in the inputs for the deploy function
 const getPrefilledDeploymentArgs = () => {
```
Comment (0xxmemo) [inline src/getDeploy/index.ts]: this doesnt look necessary to me we can just itterate over the arguments on run time of deployment function to fill the missing and or internal types

File: src/getDeploy/index.ts [hunk 4]
```diff
@@ -1,175 +1,181 @@
...
-  const rawModuleConfig = getDeploymentArgs(name, version)
-  const params = getFlattenedParams(rawModuleConfig)
+  const { deploymentArgs: rawModuleConfig } = getDeploymentConfig(name, version)
+  const params = flattenParams(rawModuleConfig)
+  const jsParams = injectJsTypes(params)
   const moduleSchema = {
-    name,
     version,
-    params,
+    params: jsParams,
```
Comment (0xxmemo) [inline src/getDeploy/index.ts]: - lets stick to the convention of calling params inputs like in the abi
- this should respect the below mentioned type ModuleSpec which extends to FormattedAbiParameter

File: src/getDeploy/index.ts [hunk 5]
```diff
@@ -1,175 +1,181 @@
...
   }
-  return moduleSchema
+  return { moduleSchema, moduleName: name }
```
Comment (0xxmemo) [inline src/getDeploy/index.ts]: why seperate the schema from the name, shouldnt the name be inside the schema?

File: src/getDeploy/index.ts [hunk 6]
```diff
@@ -1,175 +1,181 @@
...
-  return [
-    orchestratorInputs,
-    fundingManagerInputs,
-    authorizerInputs,
-    paymentProcessorInputs,
-    optionalModuleInputs,
-  ]
+  // removing orchestrator makes the rest easier because all other
+  // clientInputs can assumed to be models then
+  delete clientInputs.orchestrator
```
Comment (0xxmemo) [inline src/getDeploy/index.ts]: you can use a spread here `const { orchestrator, ...rest } = clientInputs`

File: src/getDeploy/index.ts [hunk 7]
```diff
@@ -1,175 +1,181 @@
...
-    params,
-    configData.length,
-    configData.length + dependencyData.length
-  )
-  const metadata = assembleMetadata(name, version)
-  return {
-    metadata,
-    configData: encodedConfigData,
-    dependencyData: encodedDependencyData,
+const constructArgs = (modules: any, clientInputs: any) => {
```
Comment (0xxmemo) [inline src/getDeploy/index.ts]: this should also require the `ModuleSchema` hence there wont be a need to call getDeploymentConfigs once again

( this is also another reason why the `ModuleSchema` should contain the `FormattedAbiParameter` standard, this way the arg constructor would be guided with already present data )

Comment (0xxmemo) [reply src/getDeploy/index.ts]: in this case `ModuleSchema` {} would already have the name and version fields so no need to pass modules

File: src/getDeploy/types.ts [hunk 8]
```diff
@@ -76,11 +81,8 @@ export type ModuleArg<T extends ModuleType> = {
 
 export type DeployArguments = {
   orchestrator: OrchestratorArg
```
Comment (0xxmemo) [inline src/getDeploy/types.ts:140]: this should have the same param struct as the others

**Review by 0xxmemo — COMMENTED** [review]
Overall it great that its function, but the naming and the type mismatch looks like its going to cause problems in the near feature

File: src/getDeploy/index.ts (hunk 1 above)
Comment (fabianschu) [reply src/getDeploy/index.ts]: Why do you think this is better? Imo, it is more userfriendly if we abstract away from the user our categorization of modules and distinction between mandatory and optional modules. For that reason I think it makes more sense to return to the user an object that uses the actual module names (and not the module types) as keys and that refrains from introducing the concept of optional modules. Can you elaborate on your reasoning?

Comment (0xxmemo) [reply src/getDeploy/index.ts]: tybescript doesnt infer the array type, and this helps out the internal args being more defined ( also we can rename the logicModules optionalModules ) to cover other types

**Comment by 0xxmemo** [comment]
Okay so I will share my outlook on the params here:

```ts
// New Additive type for getPossibleSpecs function
// This is a TODO, can be done after the fact- this would only be used by iterator ui's i.e ControlRoom

type PossibleModuleSpec = {
  moduleType: string // Good for type indexing / iteration 
  name: string
  version: string
}

type PossibleSpecsReturn = {
  paymentProcessor: PossibleModuleSpec[]
  fundingManager: PossibleModuleSpec[]
  authorizer: PossibleModuleSpec[]
  optionalModules: PossibleModuleSpec[]
 }
```

```ts
// Util to construct params of getDeployFunction

type GetModuleSpec<T extends ModuleType> = {
  name: string // TODO: ModuleType specific name union
  version: string // TODO: ModuleType specific version union
}
```

```ts
// Params of the getDeployFunction

type ModuleSpecs = {
  paymentProcessor: GetModuleSpec<'paymentProcessor'>
  fundingManager: GetModuleSpec<'fundingManager'>
  authorizer: GetModuleSpec<'authorizer'>
  optionalModules?: GetModuleSpec<'optionalModules'>[]
}
```

```ts
// Single type of DeploySchema

import { FormattedAbiParameter } = from 'src/types/parameter'

type ModuleSchema = {
  name: string
  version: string
  inputs: FormattedAbiParameter[]
}
```

```ts
// ReturnType of getSchema

type DeploySchema = {
  orchestrator: ModuleSchema
  paymentProcessor: ModuleSchema
  fundingManager: ModuleSchema
  authorizer: ModuleSchema
  optionalModules?: ModuleSchema[]
}
```

```ts
// Args of the deploymentFunction

type Metadata = {
    title: string
    url: string
    majorVersion: number
    minorVersion: number
  }

// Write Fn Array
type WriteFnModuleArg = {
  configData: `0x${string}`
  dependencyData: `0x${string}`
  metaData: Metadata
}

type WriteFnArgs = [
   {
    owner: `0x${string}`
    token: `0x${string}`
  },
  WriteFnModuleArg,
  WriteFnModuleArg,
  WriteFnModuleArg,
  WriteFnModuleArg[]?
]

type UserInputs<T> = any
// TODO:
// Get the User arguments per module type
type GetModuleSpecByType<
  T extends ModuleType,
  TModuleSpecs extends ModuleSpec[],
> = Extract<TModuleSpecs[number], { moduleType: T }>

type ModuleArgForType<
  T extends ModuleType,
  TModuleSpecs extends ModuleSpec[],
> = ModuleArg<
  GetModuleSpecByType<T, TModuleSpecs>['name'],
  GetModuleSpecByType<T, TModuleSpecs>['version']
>

export type DeployArguments<TModuleSpecs extends ModuleSpec[]> = {
  orchestrator: OrchestratorArg
  fundingManager: ModuleArgForType<'fundingManager', TModuleSpecs>
  authorizer: ModuleArgForType<'authorizer', TModuleSpecs>
  paymentProcessor: ModuleArgForType<'paymentProcessor', TModuleSpecs>
  logicModules?: ModuleArgForType<'logicModule', TModuleSpecs>[]
}
```

```ts
type DeploymentFunctionReturn = Promise<{
  orchestratorAddress: `0x${string}`
  txHash: `0x${string}
}>

**Pull request merged**